godot_project/project_workspace/structs/nano_particle_emitter.gd

```diff
@@ -371,16 +415,63 @@ func apply_state_snapshot(in_state_snapshot: Dictionary, in_with_instances: bool
-	if in_with_instances:
-		var group_id: int = in_state_snapshot["_instances_group"]
-		if group_id == Workspace.INVALID_STRUCTURE_ID:
-			return
+
+
+func _on_workspace_loaded(in_workspace: Workspace) -> void:
+	if _instances_group:
+		# Already initialized
+		MolecularEditorContext.workspace_loaded.disconnect(_on_workspace_loaded)
+		return
+	if in_workspace.has_structure(self):
+		initialize(in_workspace)
+		MolecularEditorContext.workspace_loaded.disconnect(_on_workspace_loaded)
+
+
+
+func _on_parameters_changed() -> void:
+	ensure_instances_exists()
+
+
+func _on_structure_reparented(in_structure: NanoStructure, in_new_parent: NanoStructure) -> void:
+	if in_structure != self or in_new_parent == _instances_group:
+		return
+	
+	_instances_atom_ids.clear()
+	_instances_bond_ids.clear()
+	_instances_group = in_new_parent
+	ensure_instances_exists.call_deferred()
+
+
+## When a simulation is applied, any atom that was emitted should remain
+## in the group and stop being tracked by this particle emitter. The other
+## atoms (still marked as invalid) should stay there.
+## This must happen before the snapshot is created.
+func _on_about_to_apply_simulation() -> void:
+	var instance_id: int = 0
+	while instance_id < _instances_atom_ids.size():
+		var first_atom_id: int = _instances_atom_ids[instance_id][0]
+		if _instances_group.is_atom_valid(first_atom_id):
+			# Instance was emitted, it's no longer handled by this emitter.
+			_instances_atom_ids.remove_at(instance_id)
+			_instances_bond_ids.remove_at(instance_id)
+		else:
+			# Instance was not emitted yet
+			instance_id += 1
+	
+	# Hidden instances are moved away before the simulation starts.
+	# Reposition them to their previous position.
+	_instances_group.start_edit()
+	for instance in _instances_atom_ids:
+		for atom_id in instance:
+			_instances_group.atom_set_position(atom_id, _instances_positions[atom_id])
+	_instances_group.end_edit()
+	
+	# Create the new instances
+	ensure_instances_exists()
+
+
+func _on_about_to_remove(in_structure: NanoStructure) -> void:
+	if in_structure == self:
+		var workspace: Workspace = MolecularEditorContext.find_workspace_possessing_structure(self)
 		var workspace_context: WorkspaceContext = MolecularEditorContext.get_workspace_context(workspace)
-		var group_structure_context: StructureContext = \
-			workspace_context.get_nano_structure_context_from_id(group_id)
-		_instances_group = group_structure_context.nano_structure as AtomicStructure
-		_instances_group.apply_state_snapshot(in_state_snapshot["_instances_group_state"])
-		group_structure_context.get_collision_engine().rebuild(group_structure_context)
-		var rendering: Rendering = workspace_context.get_rendering()
-		var renderer := rendering._get_renderer_for_atomic_structure(_instances_group)
-		renderer.apply_state_snapshot(in_state_snapshot["_instances_group_renderer_state"])
+		workspace_context.about_to_apply_simulation.disconnect(_on_about_to_apply_simulation)
```

Comment:
hummm this signal is removed on deletion, but if you undo and redo deletion i believe nothing is initializing any of the signals connected on `initialize`, at least until something calls `ensure_instances_exists`

Reply:
This was replaced with a bool flag instead (that reset on apply_snapshot). So the connection stays there, but is ignored only if the emitter is about to be deleted during a simulation.

Reply:
Signals are disconnected on remove now, and recreated each time the structure is added back to the workspace